### PR TITLE
NarrowPublicClassMethodParamTypeRule: Take default value into account

### PR DESCRIPTION
**Pull request by staabm**
even if all calling types provide the same value, when a default value of a different type is used, the rule should not suggest to change the type.

in our case the default value is intentionally not a int, as we could otherwise not differentiate it from external calls.

### CODE REVIEW

File: tests/Rules/NarrowPublicClassMethodParamTypeRule/Fixture/SkipDefaultValue.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\TypePerfect\Tests\Rules\NarrowPublicClassMethodParamTypeRule\Fixture;
+
+class SkipDefaultValue {
+    public function takesUnion(
+        int|string $value = "hallo"
```
Comment (staabm) [inline tests/Rules/NarrowPublicClassMethodParamTypeRule/Fixture/SkipDefaultValue.php]: if the default value would be an `int`, the error would be legit though

File: tests/Rules/NarrowPublicClassMethodParamTypeRule/Fixture/HandleDefaultValue.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\TypePerfect\Tests\Rules\NarrowPublicClassMethodParamTypeRule\Fixture;
+
+class HandleDefaultValue {
+    public function takesUnionWithDifferentDefaultValue(
+        int|string $value = "hallo"
```
Comment (staabm) [inline tests/Rules/NarrowPublicClassMethodParamTypeRule/Fixture/HandleDefaultValue.php:9]: should be no error here (even though all caller-types are `int` the default is a `string`)

File: tests/Rules/NarrowPublicClassMethodParamTypeRule/Fixture/HandleDefaultValue.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    ): bool
+    {
+        return true;
+    }
+
+    public function takesUnionWithSameTypeDefaultValue(
+        int|string $value = 123
```
Comment (staabm) [inline tests/Rules/NarrowPublicClassMethodParamTypeRule/Fixture/HandleDefaultValue.php:16]: here should be an error (all caller types and the default types are `int`)

**Comment by staabm** [comment]
found a fix

File: src/Collector/MethodCall/MethodCallArgTypesCollector.php [hunk 4]
```diff
@@ -34,7 +34,13 @@ public function getNodeType(): string
      */
     public function processNode(Node $node, Scope $scope): ?array
     {
-        if ($node->getArgs() === []) {
+        if ($node->getArgs() === [] || !$node->name instanceof Node\Identifier) {
```
Comment (samsonasik) [inline src/Collector/MethodCall/MethodCallArgTypesCollector.php]: when `$node->isFirstClassCallable()`, it should return null early, as `$node->getArgs()` may cause assert warning

Comment (staabm) [reply src/Collector/MethodCall/MethodCallArgTypesCollector.php]: fixed thanks (also in another spot, not related to the PR)

**Review by samsonasik — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you :clap:

**Pull request merged**